File: lib/domain/challenge/challenge_contracts.dart
import 'challenge_config.dart';
import 'challenge_evidence.dart';
import 'challenge_result.dart';
import 'challenge_type.dart';

/// Pure metadata contract defining challenge characteristics.
abstract interface class ChallengeDefinition<C extends ChallengeConfig,
    E extends ChallengeEvidence> {
  ChallengeType get type;
  String get name;
  String get family;
  int get maxDurationMs;

  ChallengeGenerator<C> get generator;
  ChallengeValidator<C, E> get validator;
  ChallengeScorer<C, E> get scorer;
}

/// Pure deterministic generator mapping seed into challenge configuration.
abstract interface class ChallengeGenerator<C extends ChallengeConfig> {
  C generate(int seed, {Map<String, dynamic>? options});
}

/// Pure validator checking evidence structural and physical integrity.
abstract interface class ChallengeValidator<C extends ChallengeConfig,
    E extends ChallengeEvidence> {
  ValidationStatus validate(C config, E evidence);
}

/// Pure mathematical scorer converting evidence to normalized 0-1000 score.
abstract interface class ChallengeScorer<C extends ChallengeConfig,
    E extends ChallengeEvidence> {
  ChallengeResult score(C config, E evidence, {ValidationStatus? statusOverride});
}

/// Abstract runtime representing the lifecycle of interactive execution state.
abstract interface class ChallengeRuntime<C extends ChallengeConfig,
    E extends ChallengeEvidence> {
  C get config;
  bool get isCompleted;
  void start();
  E complete();
}

File: supabase/functions/_shared/challenge_contracts.ts
export type ChallengeType = 'REACTION' | 'MEMORY' | 'PRECISION';

export type ValidationStatus =
  | 'VALID'
  | 'FLAGGED'
  | 'EARLY_FAULT'
  | 'TIMEOUT'
  | 'REJECTED';

export interface ChallengeConfig {
  readonly type: ChallengeType;
  readonly seed: number;
}

export interface ChallengeEvidence {
  readonly type: ChallengeType;
  readonly clientTimestampMonoMs: number;
  readonly [key: string]: unknown;
}

export interface ChallengeResult {
  readonly type: ChallengeType;
  readonly normalizedScore: number;
  readonly rawMetric: number;
  readonly formattedMetric: string;
  readonly validationStatus: ValidationStatus;
}

export interface ChallengeGenerator<TConfig extends ChallengeConfig> {
  generate(seed: number, options?: Record<string, unknown>): TConfig;
}

export interface ChallengeValidator<
  TConfig extends ChallengeConfig,
  TEvidence extends ChallengeEvidence
> {
  validate(config: TConfig, evidence: TEvidence): ValidationStatus;
}

export interface ChallengeScorer<
  TConfig extends ChallengeConfig,
  TEvidence extends ChallengeEvidence
> {
  score(config: TConfig, evidence: TEvidence): ChallengeResult;
}

export interface ChallengeDefinition<
  TConfig extends ChallengeConfig,
  TEvidence extends ChallengeEvidence
> {
  readonly type: ChallengeType;
  readonly family: string;
  readonly maxDurationMs: number;
  readonly generator: ChallengeGenerator<TConfig>;
  readonly validator: ChallengeValidator<TConfig, TEvidence>;
  readonly scorer: ChallengeScorer<TConfig, TEvidence>;
}
